src/cockpit/channels/http.py

```diff
@@ -20,145 +20,140 @@
-            return
-        finally:
-            connection.close()
-
-        self.loop.call_soon_threadsafe(self.done)
-        self.loop.call_soon_threadsafe(self.close)
-        logger.debug('closed')
-
-    def do_open(self, options):
-        logger.debug('open %s', options)
-        # TODO: generic JSON validation
-        if not options.get('method'):
-            self.close(problem='protocol-error', message='missing or empty "method" field in HTTP stream request')
-            return
-        if options.get('path') is None:
-            self.close(problem='protocol-error', message='missing "path" field in HTTP stream request')
-            return
-        if options.get('tls') is not None and options.get('unix'):
-            self.close(problem='protocol-error', message='TLS on Unix socket is not supported')
-            return
-        if options.get('connection') is not None:
-            self.close(problem='protocol-error', message='connection sharing is not implemented on this bridge')
-            return
-
-        opt_port = options.get('port')
-        opt_unix = options.get('unix')
-        if opt_port is None and opt_unix is None:
-            self.close(problem='protocol-error', message='no "port" or "unix" option for channel')
-            return
-        if opt_port is not None and opt_unix is not None:
-            self.close(problem='protocol-error', message='cannot specify both "port" and "unix" options')
-            return
+            body += data
 
-        self.options = options
-        self.body = b''
+        # Connect in a thread and handle errors
+        try:
+            await loop.run_in_executor(None, self.connect, connection, get_str(options, 'unix', None))
+        except ssl.SSLCertVerificationError as exc:
+            raise ChannelError('unknown-hostkey', message=str(exc)) from exc
+        except (OSError, IOError) as exc:
+            raise ChannelError('not-found', message=str(exc)) from exc
 
-        self.ready()
+        # Submit request in a thread and handle errors
+        try:
+            response = await loop.run_in_executor(None, self.request, connection, method, path, headers or {}, body)
+        except (http.client.HTTPException, OSError) as exc:
+            raise ChannelError('terminated', message=str(exc)) from exc
 
-    def do_data(self, data):
-        self.body += data
+        self.send_control(command='response',
+                          status=response.status,
+                          reason=response.reason,
+                          headers=self.get_headers(response, binary))
 
-    def do_done(self):
-        self.loop = asyncio.get_running_loop()
-        threading.Thread(target=self.request, daemon=True).start()
+        # Receive the body and finish up
+        try:
+            while True:
+                block = await loop.run_in_executor(None, response.read1, self.BLOCK_SIZE)
+                logger.debug('read block of size %i', len(block))
```

Comment:
TBH this log message was a debugging leftover from my initial implementation, but I suppose it doesn't  hurt much.

Reply:
I'll drop it.  I need to push again anyway with the changes to the docs and the per-channel-type json encoder.